---
title: "EDA for scRNAseq Data"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = F, cache = F)
```

This file provides exploratory data analysis for single cell RNA sequencing data using the count matrix and phenotype data as input.


```{r}
library(Biobase)
library(ggplot2)
library(ggfortify)
library(Rtsne)
library(RColorBrewer)
library(plyr)
pca.size <- 5
libsize.size <- 5
# load data
# dat is a matrix with genes as rows and cells as columns.
# pdata is a data frame with cell type specified in pdata$celltype
# nrow(pdata) == ncol(dat)
################################################################
# Insert load data code here
# dat <- dat[which(apply(dat, 1, sum)>0), ]
################################################################

```

```{r}
# calculate summary
Total.C <- colSums(dat)
Total.r <- rowSums(dat)

# useful functions
feature <- function(x) sum(x>0)

features <- function(cutoff) {
  function(x) {
    sum(x>cutoff)
  }
}

# fast pca
fast.pca <- function(data, n.pc = 2, scale = TRUE) {
  # data is sample by variable
  res <- list()
  res$center <- apply(data, 2, mean)
  res$scale <- apply(data, 2, sd)
  # we must center the data for quick calculation
  data <- sweep(data, 2, res$center, FUN = '-')
  # scale the data
  if (scale == TRUE) {
    data <- sweep(data, 2, res$scale, FUN = '/')
  }

  # svd
  # calculate the first two
  data.svd <- svd(data, nu = n.pc, nv = 0)
  res$x <- data.svd$u %*% diag(data.svd$d[1:n.pc])

  return(res)
}

```

## Basic information

The data set has `r nrow(dat)` genes and `r ncol(dat)` cells. There are `r sum(Total.r==0)` genes with 0 sequencing count for all samples. The median total counts for a gene through all samples is `r median(Total.r)`, the median average counts for a gene averaging over all samples is `r median(Total.r/ncol(dat))`. Gene expressions are observed in N cells, this N has min `r min(rowSums(dat>0))` cells, maximum `r max (rowSums(dat>0))` cells, with median `r median(rowSums(dat>0))`.

There are `r length(unique(pdata$celltype))` different cell types. A table of cell type composition is as follow.

```{r}
table(pdata$celltype)
```

The next histogram shows the histogram of all counts.

```{r}
hist(log2(dat+1), main = 'Histogram of expressions for all samples', xlab = 'log2(counts+1)', breaks=20)

hist(log2(dat), main = 'Histogram of expressions for counts>0', xlab = 'log2(counts)', breaks=20)

```

The next plot shows count distribution of all samples.

```{r}
if (max(table(pdata$celltype))<7) {
  log2.dat.1 <- log2(dat + 1)
  plt.dat <- list()
  ylim <- c(0, 0)
  xlim <- c(0, 0)
  for (i in 1:ncol(dat)) {
    hist.data <- hist(log2.dat.1[, i], breaks=40, plot=F)
    # y <- log10(hist.data$counts + 1)
    y <- hist.data$counts
    x <- hist.data$breaks
    x <- x[1:(length(x)-1)] + diff(x)/2
    plt.dat[[i]] <- data.frame(x, y)
    ylim[1] <- min(y, ylim[1])
    xlim[1] <- min(x, xlim[1])
    ylim[2] <- max(y, ylim[2])
    xlim[2] <- max(x, xlim[2])
  }
  
  
  plot(plt.dat[[1]], 
       ylab = "Frequency", 
       xlab = 'log2(Counts+1)',
       main = 'Density of Counts in Each Sample',
       ylim = ylim, xlim = xlim,
       type = 'l',
       col = factor(pdata$celltype)[1])
  for (i in 2:ncol(dat)) {
    lines(plt.dat[[i]], type = 'l',
          col = factor(pdata$celltype)[i])
  }
  legend('topright',legend = unique(pdata$celltype),
         col = unique(factor(pdata$celltype)), lty = 1)
} else {
  print('Too many samples. This plot is skipped.')
}
```


The next histogram shows the distirbution of library size across all samples.

```{r}
lib.size <- apply(dat, 2, sum)
hist(lib.size/1000, main = "", xlab = "Library Size(Thousand)", breaks = 20)
```

The boxplot shows the library size (total counts for each sample) across different cell types. Width are proportinal to the square root of the sample size.

```{r}
if (max(table(pdata$celltype))>10) {
  lib_size <- apply(dat, 2, sum)
  res_boxplot <- data.frame(celltype=pdata$celltype, lib_size)
  
  p <- ggplot(res_boxplot, aes(celltype, lib_size)) + 
    geom_boxplot(varwidth=T) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  p
  
  p <- ggplot(res_boxplot, aes(celltype, lib_size)) + 
    geom_violin() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  p
} else {
  lib_size <- apply(dat, 2, sum)
  res_boxplot <- data.frame(celltype=pdata$celltype, lib_size)
  
  p <- ggplot(res_boxplot, aes(celltype, lib_size)) + 
    geom_point() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  p
  
}

```

Histogram of distribution of mean expression for all genes. X axis is log2(gene mean expr), Y axis is log10(frequency) with frequency as in a regular histogram.

```{r}
# par(mfrow=c(1,2), mar=c(5.1,4.1,4.1,2.1), oma=c(0,0,0,0))
par(mar=c(5.1,4.1,4.1,2.1), oma=c(0,0,0,0))
hist.data = hist(log2(rowMeans(dat)), breaks=40, plot=F)
hist.data$counts = log10(hist.data$counts + 1)
plot(hist.data, 
     ylab = "log10(Frequency+1)", 
     xlab = "log2(gene mean expr)",
     main = "gene avg expr")
```

Histogram of the distribution of proportions of a gene is detected in cells. X axis is the proportion of a gene's expression > 0. Y axis is log10(frequency) with frequency as in a regular histogram.

```{r}
par(mar=c(5.1,4.1,4.1,2.1), oma=c(0,0,0,0))
hist.data = hist(rowMeans(dat>0), breaks=40, plot=F)
hist.data$counts = log10(hist.data$counts)
plot(hist.data, 
     xlab = "Proportion of a gene's expr > 0",
     ylab = "log10 Frequency",
     main = "Expressed Gene Proportion across Samples")
```


## Cell level summary
```{r}
if (ncol(dat) > 20) {
  # when the number of sample is large
  par(mfrow=c(1,3), mar=c(5.1,4.1,4.1,2.1), oma=c(0,0,0,0))
  hist(Total.C, 
       xlab = "Library size",
       main="Library size: counts across all genes", 
       breaks = 40)
  hist(apply(dat,2,feature),
       xlab="Number of expressed genes",
       main="Expressed: Gene Counts>0",
       breaks = 40)
  hist(apply(dat,2,feature)*100/nrow(dat),
       xlab="% genes expressed",
       main="% Expressed: Gene Counts>0",
       breaks = 40)
} else {
  # when the number of sample is small
  par(mfrow=c(1,3), mar=c(8.1,4.1,4.1,2.1), oma=c(0,0,0,0))
  # calculate library size
  lib.size <- apply(dat, 2, sum)
  num.exp.genes <- apply(dat, 2, function(x) sum(x>0))
  # make plots
  barplot(lib.size, 
          names.arg = names(lib.size), 
          las = 2, 
          col = mycolor,
          main="Library size of each sample"
          )
  barplot(num.exp.genes, 
          names.arg = names(lib.size), 
          las = 2, 
          col = mycolor, 
          main = "# expressed gene"
          )
  barplot(num.exp.genes/nrow(dat)*100, las = 2, col = mycolor, main = "% expressed gene")
}


```

PCA plot for each cell is as follow.

```{r, echo=FALSE}
dat.pca <- dat[which(Total.r>0), ]

# run pca
# y.pca <- prcomp(t(dat.pca), center = TRUE, scale. = TRUE)
# ylog2.pca <- prcomp(t(log2(dat.pca+1)), center = TRUE, scale. = TRUE)

y.pca <- fast.pca(t(dat.pca))
ylog2.pca <- fast.pca(t(log2(dat.pca+1)))


# plot pca
p <- autoplot(y.pca, data=pdata, colour = "celltype", size = pca.size) + 
  ggtitle("Raw Data First 2 PC")
p

p <- autoplot(ylog2.pca, data=pdata, colour = "celltype", size = pca.size) +
  ggtitle("Log2 Data First 2 PC")
p
```

tSNE plots.

```{r}
# plottSNE
# input data mat, corresponding pdata mat with celltype column
# output tSNE plot
plottSNE <- function(dat, pdata, tit, legend=F, perplexity=30) {
  # setup pdata
  pdata$celltype <- as.character(pdata$celltype)
  labels <- pdata$celltype
  pdata$idx <- 1:nrow(pdata)
  
  # colors
  colors <- brewer.pal(length(unique(pdata$celltype)),"Spectral")
  # colors <- brewer.pal(length(unique(pdata$celltype)), "BrBG")
  # colors <- c('#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99',
  #             '#e31a1c','#fdbf6f','#ff7f00','#cab2d6','#6a3d9a')
  
  colors <- data.frame(colors, 
                       celltype = unique(as.character(labels)))
  colors$colors <- as.character(colors$colors)
  
  # update pdata
  pdata <- merge(pdata, colors, by = 'celltype')
  pdata <- pdata[order(pdata$idx), ]
  
  # make plot
  tsne <- Rtsne(t(dat), dims = 2, # perplexity=perplexity, 
                verbose=F, max_iter = 1000)
  
  if (legend==T) {
    par(mar=c(5.1, 4.1, 4.1, 10.1))
    plot(tsne$Y, t='p', pch=pdata$pch, cex = 1.5,
         col = rgb(t(col2rgb(pdata$colors)/255), alpha = 0.3), 
         main = tit)
    legend(max(tsne$Y[, 1])+2, max(tsne$Y[, 2]), 
           legend = colors$celltype, 
           pch = 16, pt.cex = 1, col = colors$colors, xpd = TRUE)
  } else {
    par(mar=c(5.1, 4.1, 4.1, 2.1))
    plot(tsne$Y, t='p', pch=pdata$pch, cex = 1.5,
         col = rgb(t(col2rgb(pdata$colors)/255), alpha = 0.3), 
         main = tit)
  }
}

pdata$pch <- 16

# plottSNE(dat, pdata, "Raw Data", legend = T)
# plottSNE(log(dat+1), pdata, "log(dat+1)", legend = T)
```

We plot the histogram of the maximum counts observed for each gene across all samples.

```{r}
rmax <- apply(dat, 1, max)
hist(log2(rmax+1), breaks = 50)

hist(rmax[rmax<10], breaks = 20)
```





## Gene level summary

A gene level summary of number of cells expressed and percentage of cell expressed are summarized below. The next scatter plot shows the relationship between percentage of genes expressed and library size (Thousand).

```{r}
# colors
colors <- brewer.pal(length(unique(pdata$celltype)),"Spectral")
if (length(colors)>length(unique(pdata$celltype))) {
  colors <- colors[1:length(unique(pdata$celltype))]
}
colors <- data.frame(colors, 
                     celltype = unique(as.character(pdata$celltype)))
colors$colors <- as.character(colors$colors)
pdata$idx <- 1:nrow(pdata)
pdata <- merge(pdata, colors, by = 'celltype')
pdata <- pdata[order(pdata$idx), ]
```



```{r}
cutoffs <- seq(0, 10, 2)

for (i in 1:length(cutoffs)) {
  print(paste("Cuttoff:", cutoffs[i]))
  
  featurei <- features(cutoffs[i])
  
  par(mfrow=c(1,2), mar=c(5.1,4.1,4.1,2.1), oma=c(0,0,0,0))
  
  hist.data = hist(apply(dat,1,featurei), breaks=ncol(dat), plot=F)
  hist.data$counts = log10(hist.data$counts + 1)
  plot(hist.data, ylab = "log10 Frequency+1",
       xlab="# samples expressed",
       main = "Number of samples expressed")
  
  hist.data = hist(apply(dat,1,featurei)*100/ncol(dat), 
                   breaks=unique(apply(dat,1,featurei)*100/ncol(dat)),
                   plot=F)
  hist.data$counts = log10(hist.data$counts + 1)
  plot(hist.data, ylab = "log10 Frequency+1",
       xlab="% of samples expressed",
       main = "Percent of samples expressed")
  
  # hist(apply(dat,1,featurei),xlab="# samples expressed",
  #      breaks=ncol(dat),
  #      main="Number of samples expressed")
  # hist(apply(dat,1,featurei)*100/ncol(dat), 
  #      xlab="% of samples expressed",
  #      breaks = unique(apply(dat,1,featurei)*100/ncol(dat)), 
  # main="Percent of samples expressed")
  
  par(mfrow = c(1, 1))
  gels <- data.frame(x.ge = apply(dat,2,featurei) * 100/nrow(dat),
                     y.ls = apply(dat,2,sum)/10^3, 
                     celltype = pdata$celltype,
                     color= pdata$colors)
  p <- ggplot(gels, aes(y=x.ge, x=y.ls, color=celltype)) +
    geom_point(size = libsize.size) +
    ylab("% of genes expressed") +
    xlab("library size (Thousand)") +
    ggtitle("% Genes Expressed vs. Library Size")
  print(p)
  
}

```


```{r}
# set cutoff
cutoff <- 2
# make plot matrix
g0 <- rowMeans(dat>0)
g10 <- rowMeans(dat>cutoff)
df <- data.frame(g0, g10)
# count each unique value
plot.dat <- count(df,vars = c("g0","g10"))
# make plot
plot(plot.dat[, 1], plot.dat[, 2], 
     xlab="Proportion of gene counts>0",
     ylab=paste0("Proportion of gene counts>", cutoff), pch=16,
     cex=plot.dat[, 3]/300, col=rgb(0,0,1,alpha = 0.5))
```

The above scatter plot shows the pattern of gene expressions. Each point represents a gene. X axis represents proportion of counts>0 among cells for each gene. Y axis represents proportion of counts>`r cutoff` among cells for each gene. Points on diagnal means that for all observed genes, the observed counts are > `r cutoff`. Radius of points reflect number of overlapping points.

The following plots show the empirical CDF of cells.

```{r ,echo=FALSE}
if (ncol(dat)>30) {
  idx.ecdf <- sample(1:ncol(dat), 30)
} else {
  idx.ecdf <- 1:ncol(dat)
}
dat.ecdf <- dat[, idx.ecdf]
pdata.ecdf <- pdata[idx.ecdf, ]

# record x and y limits
x.up <- quantile(as.vector(dat.ecdf),0.999)
sa<-rep(NA,ncol(dat.ecdf))
for (i in 1:ncol(dat.ecdf)) {
  sa[i]<-ecdf(dat.ecdf[,i])(0)
}
y.low<-quantile(sa,0.01)

# make plots
# big picture
plot(ecdf(dat.ecdf[,1]),xlab="Number of gene counts",
     verticals = TRUE, 
     do.points = FALSE, ylim=c(y.low,1), xlim=c(0,x.up), 
     col = pdata.ecdf$colors[1],
     main="Cells: ECDF of gene counts")
for (i in 2:ncol(dat.ecdf)) {
  plot(ecdf(dat.ecdf[,i]),verticals = TRUE, do.points = FALSE,
       col=pdata.ecdf$colors[i], add=T)
}
legend('bottomright', legend = unique(pdata.ecdf$celltype), lty = 1, 
       col = unique(pdata.ecdf$colors), lwd = 3)

# zoomed in
x.limit <- c(0, 100)
grid <- seq(min(x.limit), max(x.limit), length.out = 500)
Fn <- ecdf(dat.ecdf[,1])
plot(grid, Fn(grid), xlab="Number of gene counts", type = 'l',
     ylim=c(y.low,1), xlim=c(0,100), col = pdata.ecdf$colors[1],
     main="Cells: ECDF of gene counts", lty = 1)
line.type <- rep(c(1, 1, 1, 2, 2, 2), 2)
for (i in 2:ncol(dat.ecdf)) {
  Fn <- ecdf(dat.ecdf[, i])
  lines(grid, Fn(grid), col = pdata.ecdf$colors[i], type = 'l')
}
legend('bottomright', legend = unique(pdata.ecdf$celltype), lty = 1, 
       col = unique(pdata.ecdf$colors), lwd = 3)

```


We summarize the genes with counts greater than 0 by showing the median and median absolute deviation (MAD) for each gene. MAD is defined as the median for the absolute deviance from the median. In the plots each point represents a gene. Bubble area reflects library size.

```{r ,echo=FALSE}
# setup the functions
cutoff <- 0
# output median after discarding values below cutoff
mid<-function(x){
  y<-x[x>cutoff]
  median(y)
}
# output MAD after discarding values below cutoff
mad<-function(x) {
  y<-x[x>cutoff]
  z<-y[y>median(y)]
  median(abs(z-median(y)))
}
y1<-apply(dat,1,mid)
y2<-apply(dat,1,mad)

# median vs MAD plot
plot(y1,y2,xlab="Median",ylab="Median absolute deviation", main=paste("Median Absolute Deviation vs Median for Genes with Counts>",cutoff,sep=""), pch=16, col=rgb(0,0,1,alpha = 0.5), cex=0.3)


# log2 data median vs MAD plot
dat.log2<-log2(dat+1)
y1.log2<-apply(dat.log2,1,mid)
y2.log2<-apply(dat.log2,1,mad)
size<-apply(dat,1,sum)
radius <- sqrt(size/ pi)

# regular plot
plot(y1.log2,y2.log2,xlab="Median",ylab="Median absolute deviation", main=paste("log2 data MAD vs Median for Gene with Counts>", cutoff, sep=""), pch=16,col=rgb(0,0,1,alpha = 0.2),cex=0.3)

# bubble plot
symbols(y1.log2, y2.log2, circles=radius,inches = 0.1, fg="white", bg="blue",xlab="Median",ylab="Median absolute deviation",main=paste("log2 data MAD vs Median for Gene with Counts>",cutoff,sep=""))

```
